### PR TITLE
protocols/autonat/tests: ignore irrelevant events

### PR DESCRIPTION
**Pull request by elenaf9**
CI failed in #2440  due to a flaky test in libp2p_autonat with "Unexpected Swarm Event". Wild guess is that a listener event happened. This PR changes it so that the autonat tests only panic on swarm-events that indicate an error in the autonat behaviour protocol, and ignore irrelevant events.

### CODE REVIEW

File: protocols/autonat/tests/test_server.rs [hunk 1]
```diff
@@ -252,8 +258,11 @@ async fn test_dial_error() {
                     break;
                 }
                 SwarmEvent::Dialing(peer) => assert_eq!(peer, client_id),
-                SwarmEvent::NewListenAddr { .. } => {}
-                _ => panic!("Unexpected Swarm Event"),
```
Comment (mxinden) [inline protocols/autonat/tests/test_server.rs:256]:
``` diff
-                 _ => panic!("Unexpected Swarm Event"),
+                 e => panic!("Unexpected Swarm Event: {:?}", e),
```

I would as well be fine to only log the error for now, thus enabling us to debug the actual issue.

**Review by mxinden — APPROVED** [review]
Alternative suggestion. I am fine either way. Your choice :)

Thanks for preparing a fix!

File: protocols/autonat/tests/test_server.rs (hunk 1 above)
Comment (elenaf9) [reply protocols/autonat/tests/test_server.rs:256]: I am bit undecided on this. But I think panicking by "blanked-implementation" could cause headaches in the future. If for example a new, unrelated swarm event is added that can happen here, the test could suddenly fail. For someone not familiar with this protocol it could take a while to figure out if the error is an actual error in the implementation, or just in the test itself because it is not up-to-date anymore. Even worse if it's a flaky error, like the one that caused #2440 to fail.

On the other hand I generally think it's good to test as much as possible, so that we also detect edge cases and unexpected interference. Considering that most swarm events don't just "randomly" occur, I am mostly concerned only about listener events. E.g. with a TCP transport a listener event is reported each time a network interface becomes available/ unavailable, which can happen at any time. 
So we could find a middle way by ignoring all listener events, and panicking on all other events like you proposed. Wdyt?

Comment (thomaseizinger) [reply protocols/autonat/tests/test_server.rs:256]: > a TCP transport a listener event is reported each time a network interface becomes available/ unavailable, which can happen at any time. So we could find a middle way by ignoring all listener events, and panicking on all other events like you proposed. Wdyt?

The test setup should be deterministic enough for this to not happen though, shouldn't it?

Comment (elenaf9) [reply protocols/autonat/tests/test_server.rs:256]: 👍 I'll change it to use the `MemoryTransport` and follow the above suggestion from @mxinden to panic in all cases and print the unexpected event.

Comment (elenaf9) [reply protocols/autonat/tests/test_server.rs:256]: Correction: The client has to listen on a TCP address. For DOS attack prevention the server only dials addresses that are bases on the observed sender IP, an refuses the dial if there is none. See https://github.com/libp2p/specs/pull/369. 
So we have to stick with the (nondeterministic) TCP transport in the tests.

**Review by mxinden — APPROVED** [review]
:pray: Thanks!

**Comment by elenaf9** [comment]
Turns out that printing the swarm event was a good idea, since the issue was is not the listener event:
https://github.com/libp2p/rust-libp2p/runs/4990919849?check_suite_focus=true. Sorry about that.
I have been running the test locally in a loop for a while now, but this error never happens there. I have the wild guess that the boot delay and/ or retry_interval is too short so that probes overlap or start while the peer is still waiting for its listening addresses.
@mxinden  I was thinking about looping the test on github actions to test it, as it was also done in https://github.com/libp2p/rust-libp2p/pull/2124, but just want double check beforehand if that's okay?

**Comment by mxinden** [comment]
> @mxinden I was thinking about looping the test on github actions to test it, as it was also done in #2124, but just want double check beforehand if that's okay?

Yes, for sure. Thanks for investigating!

---

> I have been running the test locally in a loop for a while now

Might as well be worth running single-threaded to mimic the CI environment.

https://docs.rs/async-std/latest/async_std/#runtime-configuration
